proof_system/src/sub_protocols/bbs_plus.rs

```diff
@@ -44,33 +41,69 @@ impl<'a, E: Pairing> PoKBBSSigG1SubProtocol<'a, E> {
         &mut self,
         rng: &mut R,
         blindings: BTreeMap<usize, E::ScalarField>,
-        mut witness: crate::witness::PoKBBSSignatureG1<E>,
+        witness: crate::witness::PoKBBSSignatureG1<E>,
     ) -> Result<(), ProofSystemError> {
         if self.protocol.is_some() {
             return Err(ProofSystemError::SubProtocolAlreadyInitialized(self.id));
         }
         // Create messages from revealed messages in statement and unrevealed in witness
-        let mut messages = Vec::with_capacity(self.signature_params.supported_message_count());
-        let mut revealed_indices = BTreeSet::new();
-        for i in 0..self.signature_params.supported_message_count() {
-            if witness.unrevealed_messages.contains_key(&i) {
-                messages.push(witness.unrevealed_messages.remove(&i).unwrap());
-            } else if self.revealed_messages.contains_key(&i) {
-                revealed_indices.insert(i);
-                messages.push(*self.revealed_messages.get(&i).unwrap());
-            } else {
-                return Err(ProofSystemError::BBSPlusProtocolMessageAbsent(self.id, i));
-            }
-        }
+        let mut invalid_blinding_idx = None;
+        let messages_to_commit_with_blindings = witness
+            .unrevealed_messages
+            .iter()
+            .merge_join_by(blindings, |(&m_idx, _), (b_idx, _)| m_idx.cmp(b_idx))
+            .scan((), |(), either| {
```

Comment:
Why is a scan needed here when nothing is being done with state? Shouldn't an iterator of `(usize, MessageOrBlinding)` be sufficient?

Reply:
`scan` will abort the iteration after the first `None` - we have to deal with inconsistent indices because of the current mapping from message indices to blinding.
Alternatively, we could use `.map(f).take_while(Option::is_some).flatten()`.

Reply:
Because `blindings` is built internally (by `Proof::new`) and is not accepted by the caller, we consider it safe.